tools/xkbcli-scaffold-new-layout.py

```diff
@@ -0,0 +1,323 @@
+#!/usr/bin/env python3
+#
+#  Copyright © 2020 Red Hat, Inc.
+#
+#  Permission is hereby granted, free of charge, to any person obtaining a
+#  copy of this software and associated documentation files (the "Software"),
+#  to deal in the Software without restriction, including without limitation
+#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
+#  and/or sell copies of the Software, and to permit persons to whom the
+#  Software is furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice (including the next
+#  paragraph) shall be included in all copies or substantial portions of the
+#  Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#  DEALINGS IN THE SOFTWARE.
+
+
+# This script creates the necessary scaffolding to create a custom keyboard
+# layout or option. It does not actually configure anything, it merely creates
+# the required directory structure and file scaffolding for the key
+# configurations to be added by the user.
+#
+
+import argparse
+import logging
+import os
+import re
+import sys
+
+from pathlib import Path
+from textwrap import dedent
+
+
+# Default values are set by meson but to make this usable directly within the
+# git source tree, use some sensible default values and return those where the
+# meson define hasn't been replaced.
+def default_value(key):
+    defaults = {
+        'extrapath': ('@XKBEXTRAPATH@', '/etc/xkb'),
+        'rules': ('@DEFAULT_XKB_RULES@', 'evdev'),
+    }
+
+    mesondefault, default = defaults[key]
+    if mesondefault.startswith('@') and mesondefault.endswith('@'):
+        return default
+    else:
+        return mesondefault
+
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger('xkbcli')
+logger.setLevel(logging.INFO)
+
+
+def create_directory_structure(basedir):
+    basedir.mkdir(exist_ok=True)
+    # Note: we skip geometry
+    for d in ['compat', 'keycodes', 'rules', 'symbols', 'types']:
+        (basedir / d).mkdir(exist_ok=True)
+
+
+def create_rules_template(base_path, ruleset, layout, option):
+    rules = Path(base_path) / 'rules' / ruleset
+    if rules.exists():
+        logger.warning(f'Rules file {rules} already exists, skipping')
+        return
+
+    with open(rules, 'w') as fd:
```

Comment:
`fd` for me is file-description (a number), so would just call it `f`

Reply:
one of the reasons I use `fd` here is that `file` is reserved and since 3.6 `f` is used for format strings. Like here the code would be:
```python
with open(rules, 'w') as f:
     header = dedent(f''' \
     ...
```
Which I always find a bit jarring. So `fd` is the compromise here, even though in C it's a number. This is not a hill I'm willing to die on though so I can rename it if you prefer.

Reply:
renamed to `xmlfile` and `rulesfile` in the two instances where `fd` was used previously